addons/stock/procurement.py

```diff
@@ -1,120 +1,92 @@
-        'warehouse_id': fields.many2one('stock.warehouse', 'Warehouse', help="Warehouse to consider for the route selection"),
-        'orderpoint_id': fields.many2one('stock.warehouse.orderpoint', 'Minimum Stock Rule'),
-    }
-
-    def propagate_cancels(self, cr, uid, ids, context=None):
-        move_cancel = []
-        for procurement in self.browse(cr, uid, ids, context=context):
-            if procurement.rule_id.action == 'move' and procurement.move_ids:
-                move_cancel += [m.id for m in procurement.move_ids]
-        if move_cancel:
-            self.pool.get('stock.move').action_cancel(cr, uid, move_cancel, context=context)
+
+    location_id = fields.Many2one(comodel_name='stock.location', string='Procurement Location')  # not required because task may create procurements that aren't linked to a location with sale_service
+    partner_dest_id = fields.Many2one(comodel_name='res.partner', string='Customer Address', help="In case of dropshipping, we need to know the destination address more precisely")
+    move_ids = fields.One2many('stock.move', 'procurement_id', string='Moves', help="Moves created by the procurement")
+    move_dest_id = fields.Many2one(comodel_name='stock.move', string='Destination Move', help="Move which caused (created) the procurement")
+    route_ids = fields.Many2many('stock.location.route', 'stock_location_route_procurement', 'procurement_id', 'route_id', string='Preferred Routes', help="Preferred route to be followed by the procurement order. Usually copied from the generating document (SO) but could be set up manually.")
+    warehouse_id = fields.Many2one(comodel_name='stock.warehouse', string='Warehouse', help="Warehouse to consider for the route selection")
+    orderpoint_id = fields.Many2one(comodel_name='stock.warehouse.orderpoint', string='Minimum Stock Rule')
+
+    @api.multi
+    def propagate_cancels(self):
+        self.filtered(lambda proc: proc.rule_id.action == 'move' and proc.move_ids).mapped('move_ids').action_cancel()
         return True
 
-    def cancel(self, cr, uid, ids, context=None):
-        if context is None:
-            context = {}
-        to_cancel_ids = self.get_cancel_ids(cr, uid, ids, context=context)
-        ctx = context.copy()
-        #set the context for the propagation of the procurement cancellation
-        ctx['cancel_procurement'] = True
-        self.propagate_cancels(cr, uid, to_cancel_ids, context=ctx)
-        return super(procurement_order, self).cancel(cr, uid, to_cancel_ids, context=ctx)
-
-    def _find_parent_locations(self, cr, uid, procurement, context=None):
-        location = procurement.location_id
+    @api.multi
+    def cancel(self):
+        to_cancel = self.browse(self.get_cancel_ids()).with_context({'cancel_procurement': True})
```

Comment:
You are flush the old context data which is pass calling time so please don't do this do it like : 
```
ctx = dict(self.env.context or {})
ctx['cancel_procurement'] = True
to_cancel = self.with_context(ctx).browse(self.get_cancel_ids())
```